### PR TITLE
DrivAerML: 4L/512d + gc=0.5 — transfer gc insight from AirfRANS

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

gc=0.5 just won AirfRANS (0.00277, 28.9% improvement over gc=1.0 at 4L/256d). The mechanism: sharper optimization steps with tighter gradient clipping find deeper basins in extended training. The DrivAerML baseline uses no explicit grad-clip (defaults to disabled or very high). Applying gc=0.5 to the DrivAerML 4L/512d baseline may achieve similar deep-basin exploration.

Note: Previous gc tests on DrivAerML used smaller architectures (4L/320d). gc=0.5 is being tested at the actual best architecture (4L/512d).

## Instructions

No code changes needed. Same DrivAerML baseline command but adding `--grad-clip 0.5`:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 0.5 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200 \
  --agent ymir \
  --wandb-name "ymir/drivaerml-4L512d-gc05" \
  --wandb-group "drivaerml-gc-sweep"
```

**Key things to report:**
1. Best val_primary/surface_rel_l2_pct and at which epoch
2. Does gc=0.5 improve training stability or convergence depth?
3. Grad norm behavior compared to no-clipping baseline
4. Trough envelope progression

## Baseline

- **Current best:** val_primary/surface_rel_l2_pct = **4.619%** at ep256 (PR #2691, 4L/512d + T_max=30, no explicit gc)
- **W&B run:** k8qtsxxz (267 epochs, 180-min budget)
- **External target:** <3.71% (1.24x gap)
- **Reproduce baseline:** `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT ymir:

## Results

### DrivAerML: 4L/512d + gc=0.5 (180-min budget)

| Metric | Baseline (no gc) | gc=0.5 | Delta |
|--------|------------------|--------|-------|
| **val_primary/surface_rel_l2_pct** | **4.619%** (ep256) | **6.005%** (ep160) | +30.0% (worse) |
| test_primary/surface_rel_l2_pct | N/A | 71.94% (diverged model) | N/A |
| External target | <3.71% | — | — |

**W&B run:** [nw4ywwme](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/nw4ywwme)
**Epochs:** 262 (180-min timeout), best at epoch 160
**Peak memory:** single GPU 0

### Command used
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 0.5 \
  --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 \
  --agent ymir --wandb-name "ymir/drivaerml-4L512d-gc05" --wandb-group "drivaerml-gc-sweep"
```

### What happened

**gc=0.5 does NOT transfer from AirfRANS to DrivAerML.** The hypothesis is falsified.

**Training trajectory:**
- **Phase 1 (ep1-31):** Rapid descent from 59.6% → 15.9%. Grad norms decreased from 9.5 → 1.3. Grad clipping was extremely active — 372-394 of 394 batches clipped per epoch (94-100%).
- **Phase 2 (ep31-160):** Continued steady improvement to 6.005% at ep160. Grad norms stabilized at ~0.29. Clipping events dropped dramatically to 34-69 per epoch as the model learned.
- **Phase 3 (ep161-163):** Warning signs — grad norms jumped to 0.41, clip events doubled to 76-96.
- **Phase 4 (ep164-178):** Catastrophic divergence. Grad norms exploded: 1.67 → 14 → 459 → 5246 → 344K → 8.7B → ∞. Val metric went from 6.9% to 67.2% in 15 epochs.
- **Phase 5 (ep179-262):** Complete collapse. Grad norms stuck at Infinity. Model outputs constant predictions (~67.2%), never recovered.

**Why gc=0.5 fails on DrivAerML but works on AirfRANS:**
1. **Clipping saturation:** In early training, gc=0.5 clips 94-100% of batches — far more aggressive than AirfRANS where gc=0.5 had a stabilizing effect. The DrivAerML 4L/512d model has 4x more parameters, generating larger natural gradients.
2. **Delayed divergence vs AirfRANS:** AirfRANS with T_max=5 diverged at ep205-208. DrivAerML with T_max=30 and gc=0.5 diverged at ep164 — the combination of aggressive clipping and the cosine cycling regime triggered instability sooner.
3. **No improvement over baseline:** The best gc=0.5 metric (6.005%) never came close to the baseline (4.619%). The baseline ran 267 epochs without divergence, suggesting DrivAerML's optimizer dynamics don't benefit from tighter gradient clipping.

**Trough envelope progression (at cosine troughs):**
- ep8: 31.2% → ep15: 24.9% → ep24: 19.7% → ep29: 17.1% → ep46: 10.9% → ep70: 9.3% → ep100: 7.5% → ep130: 6.7% → ep160: 6.0% → **diverged**

### Suggested follow-ups

1. **gc=1.0 or gc=1.5 at 4L/512d:** Looser clipping may provide stability benefits without the saturation problem. gc=1.0 was the AirfRANS original winner before gc=0.5. Could stabilize DrivAerML training without early divergence.
2. **gc=0.5 + reduced LR (e.g. 3e-4):** The aggressive clipping at 5e-4 may be the issue — lower LR with gc=0.5 might avoid the instability.
3. **Gradient norm warmup:** Start with no clipping (or loose clipping) for the first 50 epochs while grad norms are naturally large, then tighten to gc=0.5 for later basin exploration.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2822 because ADVISOR: Closing — your result of 6.005% does not beat the 4.619% baseline. gc=0.5 was not sufficient to prevent divergence on DrivAerML. This confirms DrivAerML needs tighter gradient control than AirfRANS. Thanks for the run, ymir..
